src/configureStore.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  createStore,
+  compose,
+  applyMiddleware,
+  combineReducers,
+  Reducer,
+  ReducersMapObject,
+  Middleware,
+  Action,
+  AnyAction,
+  StoreEnhancer,
+  Store,
+  DeepPartial
+} from 'redux'
+import { composeWithDevTools } from 'redux-devtools-extension'
+import thunk from 'redux-thunk'
+import createImmutableStateInvariantMiddleware from 'redux-immutable-state-invariant'
+import createSerializableStateInvariantMiddleware from './serializableStateInvariantMiddleware'
+
+import isPlainObject from './isPlainObject'
+
+const IS_PRODUCTION = process.env.NODE_ENV === 'production'
+
+/**
+ * Returns any array containing the default middleware installed by
+ * `configureStore()`. Useful if you want to configure your store with a custom
+ * `middleware` array but still keep the default set.
+ *
+ * @return The default middleware used by `configureStore()`.
+ */
+export function getDefaultMiddleware(
+  isProduction = IS_PRODUCTION
+): Middleware[] {
+  let middlewareArray: Middleware[] = [thunk]
+
+  if (!isProduction) {
+    middlewareArray = [
+      createImmutableStateInvariantMiddleware(),
+      thunk,
+      createSerializableStateInvariantMiddleware()
+    ]
+  }
+
+  return middlewareArray
+}
+
+/**
+ * Options for `configureStore()`.
+ */
+export interface ConfigureStoreOptions<S = any, A extends Action = AnyAction> {
+  /**
+   * A single reducer function that will be used as the root reducer, or an
+   * object of slice reducers that will be passed to `combineReducers()`.
+   */
+  reducer: Reducer<S, A> | ReducersMapObject<S, A>
+
+  /**
+   * An array of Redux middleware to install. If not supplied, defaults to
+   * the set of middleware returned by `getDefaultMiddleware()`.
+   */
+  middleware?: Middleware[]
+
+  /**
+   * Whether to enable Redux DevTools integration. Defaults to `true`.
+   */
+  devTools?: boolean
+
+  /**
+   * The initial state. You may optionally specify it to hydrate the state
+   * from the server in universal apps, or to restore a previously serialized
+   * user session. If you use `combineReducers()` to produce the root reducer
+   * function (either directly or indirectly by passing an object as `reducer`),
+   * this must be an object with the same shape as the reducer map keys.
+   */
+  preloadedState?: DeepPartial<S>
+
+  /**
+   * The store enhancers to apply. See Redux's `createStore()`. If you only
+   * need to add middleware, you can use the `middleware` parameter instaead.
+   */
+  enhancers?: StoreEnhancer[]
+}
+
+/**
+ * A friendly abstraction over the standard Redux `createStore()` function.
+ *
+ * @param config The store configuration.
+ * @returns A configured Redux store.
+ */
+export function configureStore<S = any, A extends Action = AnyAction>(
+  options: ConfigureStoreOptions<S, A>
+): Store<S, A> {
+  const {
+    reducer = undefined,
```

Comment:
It's for documentation purposes, because the type of `reducer` is actually non-nullable. There is still a "is reducer defined" runtime check for JS users (and TS users using `any` 😄).